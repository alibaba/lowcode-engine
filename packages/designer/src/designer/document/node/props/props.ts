import { computed, obx } from '@recore/obx';
import { uniqueId } from '../../../../utils/unique-id';
import { CompositeValue, PropsList, PropsMap } from '../../../schema';
import StashSpace from './stash-space';
import Prop, { IPropParent } from './prop';

export const UNSET = Symbol.for('unset');
export type UNSET = typeof UNSET;


export default class Props<O = any> implements IPropParent {
  readonly id = uniqueId('props');
  @obx.val private items: Prop[] = [];
  @obx.ref private get maps(): Map<string, Prop> {
    const maps = new Map();
    if (this.items.length > 0) {
      this.items.forEach(prop => {
        if (prop.key) {
          maps.set(prop.key, prop);
        }
      });
    }
    return maps;
  }

  private stash = new StashSpace(
    prop => {
      this.items.push(prop);
      prop.parent = this;
    },
    () => {
      return true;
    },
  );

  /**
   * 元素个数
   */
  get size() {
    return this.items.length;
  }

  @computed get value(): PropsMap | PropsList | null {
    if (this.items.length < 1) {
      return null;
    }
    if (this.type === 'list') {
      return this.items.map(item => ({
        spread: item.spread,
        name: item.key as string,
        value: item.value,
      }));
    }
    const maps: any = {};
    this.items.forEach(prop => {
      if (prop.key) {
        maps[prop.key] = prop.value;
      }
    });
    return maps;
  }

  @obx type: 'map' | 'list' = 'map';

  constructor(readonly owner: O, value?: PropsMap | PropsList | null) {
    if (Array.isArray(value)) {
      this.type = 'list';
      this.items = value.map(item => new Prop(this, item.value, item.name, item.spread));
    } else if (value != null) {
      this.type = 'map';
      this.items = Object.keys(value).map(key => new Prop(this, value[key], key));
    }
  }

  /**
   * 根据 path 路径查询属性，如果没有则临时生成一个
   */
  query(path: string): Prop;
  /**
   * 根据 path 路径查询属性
   *
   * @useStash 如果没有则临时生成一个
   */
  query(path: string, useStash: true): Prop;
  /**
   * 根据 path 路径查询属性
   */
  query(path: string, useStash: false): Prop | null;
  /**
   * 根据 path 路径查询属性
   *
   * @useStash 如果没有则临时生成一个
   */
  query(path: string, useStash: boolean = true) {
    let matchedLength = 0;
    let firstMatched = null;
    if (this.items) {
      // target: a.b.c
      // trys: a.b.c, a.b, a
      let i = this.items.length;
      while (i-- > 0) {
        const expr = this.items[i];
        if (!expr.key) {
          continue;
        }
        const name = String(expr.key);
        if (name === path) {
          // completely match
          return expr;
        }

        // fisrt match
        const l = name.length;
        if (path.slice(0, l + 1) === `${name}.`) {
          matchedLength = l;
          firstMatched = expr;
        }
      }
    }

    let ret = null;
    if (firstMatched) {
      ret = firstMatched.get(path.slice(matchedLength + 1), true);
    }
    if (!ret && useStash) {
      return this.stash.get(path);
    }

    return ret;
  }

  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   * @param useStash 强制
   */
  get(path: string, useStash: true): Prop;
  /**
   * 获取某个属性
   * @param useStash 强制
   */
  get(path: string, useStash: false): Prop | null;
  /**
   * 获取某个属性
   */
  get(path: string): Prop | null;
  get(name: string, useStash = false) {
    return this.maps.get(name) || (useStash && this.stash.get(name)) || null;
  }

  /**
   * 删除项
   */
  delete(prop: Prop): void {
    const i = this.items.indexOf(prop);
    if (i > -1) {
      this.items.splice(i, 1);
      prop.purge();
    }
  }

  /**
   * 删除 key
   */
  deleteKey(key: string): void {
    this.items = this.items.filter(item => {
      if (item.key === key) {
        item.purge();
        return false;
      }
      return true;
    });
  }

  /**
   * 添加值
   */
  add(value: CompositeValue | null, key?: string | number, spread = false): Prop {
    const prop = new Prop(this, value, key, spread);
    this.items.push(prop);
    return prop;
  }

  /**
   * 是否存在 key
   */
  has(key: string): boolean {
    return this.maps.has(key);
  }

  /**
   * 迭代器
   */
  [Symbol.iterator](): { next(): { value: Prop } } {
    let index = 0;
    const items = this.items;
    const length = items.length || 0;
    return {
      next() {
        if (index < length) {
          return {
            value: items[index++],
            done: false,
          };
        }
        return {
          value: undefined as any,
          done: true,
        };
      },
    };
  }

  /**
   * 遍历
   */
  forEach(fn: (item: Prop, key: number | string | undefined) => void): void {
    this.items.forEach(item => {
      return fn(item, item.key);
    });
  }

  /**
   * 遍历
   */
  map<T>(fn: (item: Prop, key: number | string | undefined) => T): T[] | null {
    return this.items.map(item => {
      return fn(item, item.key);
    });
  }

  private purged = false;
  /**
   * 回收销毁
   */
  purge() {
    if (this.purged) {
      return;
    }
    this.purged = true;
    this.stash.purge();
    this.items.forEach(item => item.purge());
  }
}
