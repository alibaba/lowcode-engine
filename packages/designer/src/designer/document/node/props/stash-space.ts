import { obx, autorun, untracked } from '@recore/obx';
import Prop, { IPropParent } from './prop';

export type PendingItem = Prop[];
export default class StashSpace implements IPropParent {
  @obx.val private space: Set<Prop> = new Set();
  @obx.ref private get maps(): Map<string, Prop> {
    const maps = new Map();
    if (this.space.size > 0) {
      this.space.forEach(prop => {
        maps.set(prop.key, prop);
      });
    }
    return maps;
  }
  private willPurge: () => void;

  constructor(write: (item: Prop) => void, before: () => boolean) {
    this.willPurge = autorun(() => {
      if (this.space.size < 1) {
        return;
      }
      const pending: Prop[] = [];
      for (const prop of this.space) {
        if (!prop.isUnset()) {
          this.space.delete(prop);
          pending.push(prop);
        }
      }
      if (pending.length > 0) {
        untracked(() => {
          if (before()) {
            for (const item of pending) {
              write(item);
            }
          }
        });
      }
    });
  }

  get(key: string): Prop {
    let prop = this.maps.get(key);
    if (!prop) {
      prop = new Prop(this, null, key);
      this.space.add(prop);
    }
    return prop;
  }

  delete(prop: Prop) {
    this.space.delete(prop);
    prop.purge();
  }

  clear() {
    this.space.forEach(item => item.purge());
    this.space.clear();
  }

  purge() {
    this.willPurge();
    this.space.clear();
  }
}
