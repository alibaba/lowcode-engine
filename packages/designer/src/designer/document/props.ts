import { untracked, computed, obx } from '@recore/obx';
import { uniqueId, isPlainObject, hasOwnProperty } from '../../utils';
import { valueToSource } from '../../utils/value-to-source';
import { CompositeValue, isJSExpression, PropsList, PropsMap } from '../schema';
import StashSpace from './stash-space';

export const UNSET = Symbol.for('unset');
export type UNSET = typeof UNSET;

export interface IPropParent {
  delete(prop: Prop): void;
}

export class Prop implements IPropParent {
  readonly isProp = true;

  readonly id = uniqueId('prop$');

  private _type: 'unset' | 'literal' | 'map' | 'list' | 'expression' = 'unset';
  /**
   * 属性类型
   */
  get type(): 'unset' | 'literal' | 'map' | 'list' | 'expression' {
    return this._type;
  }

  @obx.ref private _value: any = UNSET;

  /**
   * 属性值
   */
  @computed get value(): CompositeValue {
    if (this._type === 'unset') {
      return null;
    }

    const type = this._type;
    if (type === 'literal' || type === 'expression') {
      return this._value;
    }

    if (type === 'map') {
      if (!this._items) {
        return this._value;
      }
      const maps: any = {};
      this.items!.forEach((prop, key) => {
        maps[key] = prop.value;
      });
      return maps;
    }

    if (type === 'list') {
      if (!this._items) {
        return this._items;
      }
      return this.items!.map(prop => prop.value);
    }

    return null;
  }

  /**
   * set value, val should be JSON Object
   */
  set value(val: CompositeValue) {
    this._value = val;
    const t = typeof val;
    if (val == null) {
      this._value = null;
      this._type = 'literal';
    } else if (t === 'string' || t === 'number' || t === 'boolean') {
      this._value = val;
      this._type = 'literal';
    } else if (Array.isArray(val)) {
      this._type = 'list';
    } else if (isPlainObject(val)) {
      if (isJSExpression(val)) {
        this._type = 'expression';
      } else {
        this._type = 'map';
      }
      this._type = 'map';
    } else {
      this._type = 'expression';
      this._value = {
        type: 'JSExpression',
        value: valueToSource(val),
      };
    }
    if (untracked(() => this._items)) {
      this._items!.forEach(prop => prop.purge());
      this._items = null;
    }
    this._maps = null;
    if (this.stash) {
      this.stash.clear();
    }
  }

  /**
   * 取消设置值
   */
  unset() {
    this._type = 'unset';
  }

  /**
   * 是否未设置值
   */
  isUnset() {
    return this._type === 'unset';
  }

  /**
   * 值是否包含表达式
   * 包含 JSExpresion | JSSlot 等值
   */
  isContainJSExpression(): boolean {
    const type = this._type;
    if (type === 'expression') {
      return true;
    }
    if (type === 'literal' || type === 'unset') {
      return false;
    }
    if ((type === 'list' || type === 'map') && this.items) {
      return this.items.some(item => item.isContainJSExpression());
    }
    return false;
  }

  /**
   * 是否简单 JSON 数据
   */
  isJSON() {
    return !this.isContainJSExpression();
  }

  private _items: Prop[] | null = null;
  private _maps: Map<string, Prop> | null = null;
  @computed private get items(): Prop[] | null {
    let _items: any;
    untracked(() => {
      _items = this._items;
    });
    if (!_items) {
      if (this._type === 'list') {
        const data = this._value;
        const items = [];
        for (const item of data) {
          items.push(new Prop(this, item));
        }
        _items = items;
        this._maps = null;
      } else if (this._type === 'map') {
        const data = this._value;
        const items = [];
        const maps = new Map<string, Prop>();
        const keys = Object.keys(data);
        for (const key of keys) {
          const prop = new Prop(this, data[key], key);
          items.push(prop);
          maps.set(key, prop);
        }
        _items = items;
        this._maps = maps;
      } else {
        _items = null;
        this._maps = null;
      }
      this._items = _items;
    }
    return _items;
  }
  @computed private get maps(): Map<string, Prop> | null {
    if (!this.items || this.items.length < 1) {
      return null;
    }
    return this._maps;
  }

  private stash: StashSpace | undefined;

  /**
   * 键值
   */
  @obx key: string | number | undefined;
  /**
   * 扩展值
   */
  @obx spread: boolean;

  constructor(
    public parent: IPropParent,
    value: CompositeValue | UNSET = UNSET,
    key?: string | number,
    spread = false,
  ) {
    if (value !== UNSET) {
      this.value = value;
    }
    this.key = key;
    this.spread = spread;
  }

  /**
   * 获取某个属性
   * @param stash 强制
   */
  get(path: string, stash: false): Prop | null;
  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   * @param stash 强制
   */
  get(path: string, stash: true): Prop;
  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   */
  get(path: string): Prop;
  get(path: string, stash = true) {
    const type = this._type;
    if (type !== 'map' && type !== 'unset' && !stash) {
      return null;
    }

    const maps = type === 'map' ? this.maps : null;

    let prop: any = maps ? maps.get(path) : null;

    if (prop) {
      return prop;
    }

    const i = path.indexOf('.');
    let entry = path;
    let nest = '';
    if (i > 0) {
      nest = path.slice(i + 1);
      if (nest) {
        entry = path.slice(0, i);
        prop = maps ? maps.get(entry) : null;
        if (prop) {
          return prop.get(nest, stash);
        }
      }
    }

    if (stash) {
      if (!this.stash) {
        this.stash = new StashSpace(
          item => {
            // item take effect
            this.set(String(item.key), item);
            item.parent = this;
          },
          () => {
            return true;
          },
        );
      }
      prop = this.stash.get(entry);
      if (nest) {
        return prop.get(nest, true);
      }

      return prop;
    }

    return null;
  }

  /**
   * 从父级移除本身
   */
  remove() {
    this.parent.delete(this);
  }

  /**
   * 删除项
   */
  delete(prop: Prop): void {
    if (this.items) {
      const i = this.items.indexOf(prop);
      if (i > -1) {
        this.items.slice(i, 1);
        prop.purge();
      }
      if (this._maps && prop.key) {
        this._maps.delete(String(prop.key));
      }
    }
  }

  /**
   * 删除 key
   */
  deleteKey(key: string): void {
    if (this.maps) {
      const prop = this.maps.get(key);
      if (prop) {
        this.delete(prop);
      }
    }
  }

  /**
   * 元素个数
   */
  size(): number {
    return this.items?.length || 0;
  }

  /**
   * 添加值到列表
   *
   * @param force 强制
   */
  add(value: CompositeValue, force = false): Prop | null {
    const type = this._type;
    if (type !== 'list' && type !== 'unset' && !force) {
      return null;
    }
    if (type === 'unset' || (force && type !== 'list')) {
      this.value = [];
    }
    const prop = new Prop(this, value);
    this.items!.push(prop);
    return prop;
  }

  /**
   * 设置值到字典
   *
   * @param force 强制
   */
  set(key: string, value: CompositeValue | Prop, force = false) {
    const type = this._type;
    if (type !== 'map' && type !== 'unset' && !force) {
      return null;
    }
    if (type === 'unset' || (force && type !== 'map')) {
      this.value = {};
    }
    const prop = isProp(value) ? value : new Prop(this, value, key);
    const items = this.items!;
    const maps = this.maps!;
    const orig = maps.get(key);
    if (orig) {
      // replace
      const i = items.indexOf(orig);
      if (i > -1) {
        items.splice(i, 1, prop)[0].purge();
      }
      maps.set(key, prop);
    } else {
      // push
      items.push(prop);
      maps.set(key, prop);
    }

    return prop;
  }

  /**
   * 是否存在 key
   */
  has(key: string): boolean {
    if (this._type !== 'map') {
      return false;
    }
    if (this._maps) {
      return this._maps.has(key);
    }
    return hasOwnProperty(this._value, key);
  }

  private purged = false;
  /**
   * 回收销毁
   */
  purge() {
    if (this.purged) {
      return;
    }
    this.purged = true;
    if (this.stash) {
      this.stash.purge();
    }
    if (this._items) {
      this._items.forEach(item => item.purge());
    }
    this._maps = null;
  }

  /**
   * 迭代器
   */
  [Symbol.iterator](): { next(): { value: Prop } } {
    let index = 0;
    const items = this.items;
    const length = items?.length || 0;
    return {
      next() {
        if (index < length) {
          return {
            value: items![index++],
            done: false,
          };
        }
        return {
          value: undefined as any,
          done: true,
        };
      },
    };
  }

  /**
   * 遍历
   */
  forEach(fn: (item: Prop, key: number | string | undefined) => void): void {
    const items = this.items;
    if (!items) {
      return;
    }
    const isMap = this._type === 'map';
    items.forEach((item, index) => {
      return isMap ? fn(item, item.key) : fn(item, index);
    });
  }

  /**
   * 遍历
   */
  map<T>(fn: (item: Prop, key: number | string | undefined) => T): T[] | null {
    const items = this.items;
    if (!items) {
      return null;
    }
    const isMap = this._type === 'map';
    return items.map((item, index) => {
      return isMap ? fn(item, item.key) : fn(item, index);
    });
  }
}

export function isProp(obj: any): obj is Prop {
  return obj && obj.isProp;
}

export default class Props<O = any> implements IPropParent {
  @obx.val private items: Prop[] = [];
  @obx.ref private get maps(): Map<string, Prop> {
    const maps = new Map();
    if (this.items.length > 0) {
      this.items.forEach(prop => {
        if (prop.key) {
          maps.set(prop.key, prop);
        }
      });
    }
    return maps;
  }

  private stash = new StashSpace(
    prop => {
      this.items.push(prop);
      prop.parent = this;
    },
    () => {
      return true;
    },
  );

  /**
   * 元素个数
   */
  get size() {
    return this.items.length;
  }

  @computed get value(): PropsMap | PropsList | null {
    if (this.items.length < 1) {
      return null;
    }
    if (this.type === 'list') {
      return this.items.map(item => ({
        spread: item.spread,
        name: item.key as string,
        value: item.value,
      }));
    }
    const maps: any = {};
    this.items.forEach(prop => {
      if (prop.key) {
        maps[prop.key] = prop.value;
      }
    });
    return maps;
  }

  @obx type: 'map' | 'list' = 'map';

  constructor(readonly owner: O, value?: PropsMap | PropsList | null) {
    if (Array.isArray(value)) {
      this.type = 'list';
      value.forEach(item => {});
    } else if (value != null) {
      this.type = 'map';
    }
  }

  /**
   * 根据 path 路径查询属性，如果没有则临时生成一个
   */
  query(path: string): Prop;
  /**
   * 根据 path 路径查询属性
   *
   * @useStash 如果没有则临时生成一个
   */
  query(path: string, useStash: true): Prop;
  /**
   * 根据 path 路径查询属性
   */
  query(path: string, useStash: false): Prop | null;
  /**
   * 根据 path 路径查询属性
   *
   * @useStash 如果没有则临时生成一个
   */
  query(path: string, useStash: boolean = true) {
    let matchedLength = 0;
    let firstMatched = null;
    if (this.items) {
      // target: a.b.c
      // trys: a.b.c, a.b, a
      let i = this.items.length;
      while (i-- > 0) {
        const expr = this.items[i];
        if (!expr.key) {
          continue;
        }
        const name = String(expr.key);
        if (name === path) {
          // completely match
          return expr;
        }

        // fisrt match
        const l = name.length;
        if (path.slice(0, l + 1) === `${name}.`) {
          matchedLength = l;
          firstMatched = expr;
        }
      }
    }

    let ret = null;
    if (firstMatched) {
      ret = firstMatched.get(path.slice(matchedLength + 1), true);
    }
    if (!ret && useStash) {
      return this.stash.get(path);
    }

    return ret;
  }

  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   * @param useStash 强制
   */
  get(path: string, useStash: true): Prop;
  /**
   * 获取某个属性
   * @param useStash 强制
   */
  get(path: string, useStash: false): Prop | null;
  /**
   * 获取某个属性
   */
  get(path: string): Prop | null;
  get(name: string, useStash = false) {
    return this.maps.get(name) || (useStash && this.stash.get(name)) || null;
  }

  /**
   * 删除项
   */
  delete(prop: Prop): void {
    const i = this.items.indexOf(prop);
    if (i > -1) {
      this.items.splice(i, 1);
      prop.purge();
    }
  }

  /**
   * 删除 key
   */
  deleteKey(key: string): void {
    this.items = this.items.filter(item => {
      if (item.key === key) {
        item.purge();
        return false;
      }
      return true;
    });
  }

  /**
   * 添加值
   */
  add(value: CompositeValue | null, key?: string | number, spread = false): Prop {
    const prop = new Prop(this, value, key, spread);
    this.items.push(prop);
    return prop;
  }

  /**
   * 是否存在 key
   */
  has(key: string): boolean {
    return this.maps.has(key);
  }

  /**
   * 迭代器
   */
  [Symbol.iterator](): { next(): { value: Prop } } {
    let index = 0;
    const items = this.items;
    const length = items.length || 0;
    return {
      next() {
        if (index < length) {
          return {
            value: items[index++],
            done: false,
          };
        }
        return {
          value: undefined as any,
          done: true,
        };
      },
    };
  }

  /**
   * 遍历
   */
  forEach(fn: (item: Prop, key: number | string | undefined) => void): void {
    this.items.forEach(item => {
      return fn(item, item.key);
    });
  }

  /**
   * 遍历
   */
  map<T>(fn: (item: Prop, key: number | string | undefined) => T): T[] | null {
    return this.items.map(item => {
      return fn(item, item.key);
    });
  }

  private purged = false;
  /**
   * 回收销毁
   */
  purge() {
    if (this.purged) {
      return;
    }
    this.purged = true;
    this.stash.purge();
    this.items.forEach(item => item.purge());
  }
}
