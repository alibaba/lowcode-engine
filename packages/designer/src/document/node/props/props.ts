import { computed, makeObservable, obx, action } from '@alilc/lowcode-editor-core';
import { PropsMap, PropsList, CompositeValue } from '@alilc/lowcode-types';
import { uniqueId, compatStage } from '@alilc/lowcode-utils';
import { Prop, IPropParent, UNSET } from './prop';
import { Node } from '../node';
import { TransformStage } from '../transform-stage';

interface ExtrasObject {
  [key: string]: any;
}

export const EXTRA_KEY_PREFIX = '___';
export function getConvertedExtraKey(key: string): string {
  if (!key) {
    return '';
  }
  let _key = key;
  if (key.indexOf('.') > 0) {
    _key = key.split('.')[0];
  }
  return EXTRA_KEY_PREFIX + _key + EXTRA_KEY_PREFIX + key.slice(_key.length);
}
export function getOriginalExtraKey(key: string): string {
  return key.replace(new RegExp(`${EXTRA_KEY_PREFIX}`, 'g'), '');
}
export class Props implements IPropParent {
  readonly id = uniqueId('props');

  @obx.shallow private items: Prop[] = [];

  @computed private get maps(): Map<string, Prop> {
    const maps = new Map();
    if (this.items.length > 0) {
      this.items.forEach((prop) => {
        if (prop.key) {
          maps.set(prop.key, prop);
        }
      });
    }
    return maps;
  }

  readonly path = [];

  get props(): Props {
    return this;
  }

  readonly owner: Node;

  /**
   * 元素个数
   */
  @computed get size() {
    return this.items.length;
  }

  @obx type: 'map' | 'list' = 'map';

  constructor(owner: Node, value?: PropsMap | PropsList | null, extras?: ExtrasObject) {
    makeObservable(this);
    this.owner = owner;
    if (Array.isArray(value)) {
      this.type = 'list';
      this.items = value.map(
        (item, idx) => new Prop(this, item.value, item.name || idx, item.spread),
      );
    } else if (value != null) {
      this.items = Object.keys(value).map((key) => new Prop(this, value[key], key, false));
    }
    if (extras) {
      Object.keys(extras).forEach((key) => {
        this.items.push(new Prop(this, (extras as any)[key], getConvertedExtraKey(key)));
      });
    }
  }

  @action
  import(value?: PropsMap | PropsList | null, extras?: ExtrasObject) {
    const originItems = this.items;
    if (Array.isArray(value)) {
      this.type = 'list';
      this.items = value.map(
        (item, idx) => new Prop(this, item.value, item.name || idx, item.spread),
      );
    } else if (value != null) {
      this.type = 'map';
      this.items = Object.keys(value).map((key) => new Prop(this, value[key], key));
    } else {
      this.type = 'map';
      this.items = [];
    }
    if (extras) {
      Object.keys(extras).forEach((key) => {
        this.items.push(new Prop(this, (extras as any)[key], getConvertedExtraKey(key)));
      });
    }
    originItems.forEach((item) => item.purge());
  }

  @action
  merge(value: PropsMap, extras?: PropsMap) {
    Object.keys(value).forEach((key) => {
      this.query(key, true)!.setValue(value[key]);
      this.query(key, true)!.setupItems();
    });
    if (extras) {
      Object.keys(extras).forEach((key) => {
        this.query(getConvertedExtraKey(key), true)!.setValue(extras[key]);
        this.query(getConvertedExtraKey(key), true)!.setupItems();
      });
    }
  }

  export(stage: TransformStage = TransformStage.Save): {
    props?: PropsMap | PropsList;
    extras?: ExtrasObject;
  } {
    stage = compatStage(stage);
    if (this.items.length < 1) {
      return {};
    }
    let allProps = {} as any;
    let props: any = {};
    const extras: any = {};
    if (this.type === 'list') {
      props = [];
      this.items.forEach((item) => {
        let value = item.export(stage);
        let name = item.key as string;
        if (name && typeof name === 'string' && name.startsWith(EXTRA_KEY_PREFIX)) {
          name = getOriginalExtraKey(name);
          extras[name] = value;
        } else {
          props.push({
            spread: item.spread,
            name,
            value,
          });
        }
      });
    } else {
      this.items.forEach((item) => {
        let name = item.key as string;
        if (name == null || item.isUnset() || item.isVirtual()) return;
        let value = item.export(stage);
        if (value != null) {
          allProps[name] = value;
        }
      });
      // compatible vision
      const transformedProps = this.transformToStatic(allProps);
      Object.keys(transformedProps).forEach((name) => {
        const value = transformedProps[name];
        if (typeof name === 'string' && name.startsWith(EXTRA_KEY_PREFIX)) {
          name = getOriginalExtraKey(name);
          extras[name] = value;
        } else {
          props[name] = value;
        }
      });
    }

    return { props, extras };
  }

  /**
   * @deprecated
   */
  /* istanbul ignore next */
  private transformToStatic(props: any) {
    let transducers = this.owner.componentMeta?.prototype?.options?.transducers;
    if (!transducers) {
      return props;
    }
    if (!Array.isArray(transducers)) {
      transducers = [transducers];
    }
    props = transducers.reduce((xprops: any, transducer: any) => {
      if (transducer && typeof transducer.toStatic === 'function') {
        return transducer.toStatic(xprops);
      }
      return xprops;
    }, props);
    return props;
  }

  /**
   * 根据 path 路径查询属性
   *
   * @param createIfNone 当没有的时候，是否创建一个
   */
  @action
  query(path: string, createIfNone = true): Prop | null {
    return this.get(path, createIfNone);
  }

  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   * @param createIfNone 当没有的时候，是否创建一个
   */
  @action
  get(path: string, createIfNone = false): Prop | null {
    let entry = path;
    let nest = '';
    const i = path.indexOf('.');
    if (i > 0) {
      nest = path.slice(i + 1);
      if (nest) {
        entry = path.slice(0, i);
      }
    }

    let prop = this.maps.get(entry);
    if (!prop && createIfNone) {
      prop = new Prop(this, UNSET, entry);
      this.items.push(prop);
    }

    if (prop) {
      return nest ? prop.get(nest, createIfNone) : prop;
    }

    return null;
  }

  /**
   * 删除项
   */
  @action
  delete(prop: Prop): void {
    const i = this.items.indexOf(prop);
    if (i > -1) {
      this.items.splice(i, 1);
      prop.purge();
    }
  }

  /**
   * 删除 key
   */
  @action
  deleteKey(key: string): void {
    this.items = this.items.filter((item, i) => {
      if (item.key === key) {
        item.purge();
        this.items.splice(i, 1);
        return false;
      }
      return true;
    });
  }

  /**
   * 添加值
   */
  @action
  add(
    value: CompositeValue | null,
    key?: string | number,
    spread = false,
    options: any = {},
  ): Prop {
    const prop = new Prop(this, value, key, spread, options);
    this.items.push(prop);
    return prop;
  }

  /**
   * 是否存在 key
   */
  has(key: string): boolean {
    return this.maps.has(key);
  }

  /**
   * 迭代器
   */
  [Symbol.iterator](): { next(): { value: Prop } } {
    let index = 0;
    const { items } = this;
    const length = items.length || 0;
    return {
      next() {
        if (index < length) {
          return {
            value: items[index++],
            done: false,
          };
        }
        return {
          value: undefined as any,
          done: true,
        };
      },
    };
  }

  /**
   * 遍历
   */
  @action
  forEach(fn: (item: Prop, key: number | string | undefined) => void): void {
    this.items.forEach((item) => {
      return fn(item, item.key);
    });
  }

  /**
   * 遍历
   */
  @action
  map<T>(fn: (item: Prop, key: number | string | undefined) => T): T[] | null {
    return this.items.map((item) => {
      return fn(item, item.key);
    });
  }

  @action
  filter(fn: (item: Prop, key: number | string | undefined) => boolean) {
    return this.items.filter((item) => {
      return fn(item, item.key);
    });
  }

  private purged = false;

  /**
   * 回收销毁
   */
  @action
  purge() {
    if (this.purged) {
      return;
    }
    this.purged = true;
    this.items.forEach((item) => item.purge());
  }

  /**
   * 获取某个属性, 如果不存在，临时获取一个待写入
   * @param createIfNone 当没有的时候，是否创建一个
   */
  @action
  getProp(path: string, createIfNone = true): Prop | null {
    return this.query(path, createIfNone) || null;
  }

  /**
   * 获取单个属性值
   */
  @action
  getPropValue(path: string): any {
    return this.getProp(path, false)?.value;
  }

  /**
   * 设置单个属性值
   */
  @action
  setPropValue(path: string, value: any) {
    this.getProp(path, true)!.setValue(value);
  }

  /**
   * 获取 props 对应的 node
   */
  getNode() {
    return this.owner;
  }

  /**
   * @deprecated
   * 获取 props 对应的 node
   */
  @action
  toData() {
    return this.export()?.props;
  }
}
