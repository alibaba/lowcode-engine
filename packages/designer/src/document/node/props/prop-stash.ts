import { obx, autorun, untracked, computed, makeObservable, action } from '@ali/lowcode-editor-core';
import { Prop, IPropParent, UNSET } from './prop';
import { Props } from './props';
import { Node } from '../node';

export type PendingItem = Prop[];
export class PropStash implements IPropParent {
  readonly isPropStash = true;

  @obx.shallow private space: Set<Prop> = new Set();

  @computed private get maps(): Map<string | number, Prop> {
    const maps = new Map();
    if (this.space.size > 0) {
      this.space.forEach(prop => {
        maps.set(prop.key, prop);
      });
    }
    return maps;
  }

  private willPurge: () => void;

  readonly owner: Node;

  constructor(readonly props: Props, write: (item: Prop) => void) {
    makeObservable(this);
    this.owner = props.owner;
    this.willPurge = autorun(() => {
      if (this.space.size < 1) {
        return;
      }
      const pending: Prop[] = [];
      for (const prop of this.space) {
        if (!prop.isUnset() && !prop.isVirtual()) {
          this.space.delete(prop);
          pending.push(prop);
        }
      }
      if (pending.length > 0) {
        untracked(() => {
          for (const item of pending) {
            write(item);
          }
        });
      }
    });
  }

  @action
  get(key: string | number): Prop {
    let prop = this.maps.get(key);
    if (!prop) {
      prop = new Prop(this, UNSET, key);
      this.space.add(prop);
    }
    return prop;
  }

  @action
  delete(prop: Prop) {
    this.space.delete(prop);
    prop.purge();
  }

  @action
  clear() {
    this.space.forEach(item => item.purge());
    this.space.clear();
  }

  @action
  purge() {
    this.willPurge();
    this.space.clear();
  }
}
