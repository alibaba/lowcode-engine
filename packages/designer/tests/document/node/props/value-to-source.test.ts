// @ts-nocheck
import '../../../fixtures/silent-console';
import { getSource, valueToSource } from '../../../../src/document/node/props/value-to-source';

it('valueToSource', () => {
  expect(valueToSource(1)).toMatchSnapshot();
  expect(valueToSource(true)).toMatchSnapshot();
  expect(valueToSource([])).toMatchSnapshot();
  expect(valueToSource([{ a: 1 }])).toMatchSnapshot();
  expect(valueToSource({ a: 1 })).toMatchSnapshot();
  expect(valueToSource(null)).toMatchSnapshot();
  expect(valueToSource(() => {})).toMatchSnapshot();
  expect(valueToSource(new Map())).toMatchSnapshot();
  expect(valueToSource(new Set())).toMatchSnapshot();
  expect(valueToSource(/haha/)).toMatchSnapshot();
  expect(valueToSource('hahah')).toMatchSnapshot();
  expect(valueToSource(Symbol('haha'))).toMatchSnapshot();
  expect(valueToSource()).toMatchSnapshot();
  expect(valueToSource(new Date(1607680998520))).toMatchSnapshot();
});

it('getSource', () => {
  expect(getSource({ __source: { a: 1 } })).toEqual({ a: 1 });
  expect(getSource()).toBe('');
  const value = { abc: 1 };
  getSource(value);
  expect(value).toHaveProperty('__source');
  expect(getSource(1)).toBe('1');
});
