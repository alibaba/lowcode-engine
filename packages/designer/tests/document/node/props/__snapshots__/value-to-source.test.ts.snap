// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valueToSource 1`] = `"1"`;

exports[`valueToSource 2`] = `"true"`;

exports[`valueToSource 3`] = `"[]"`;

exports[`valueToSource 4`] = `
"[{
  \\"a\\": 1
}]"
`;

exports[`valueToSource 5`] = `
"{
  \\"a\\": 1
}"
`;

exports[`valueToSource 6`] = `"null"`;

exports[`valueToSource 7`] = `"() => {}"`;

exports[`valueToSource 8`] = `"new Map()"`;

exports[`valueToSource 9`] = `"new Set()"`;

exports[`valueToSource 10`] = `"/haha/"`;

exports[`valueToSource 11`] = `"\\"hahah\\""`;

exports[`valueToSource 12`] = `"Symbol(\\"haha\\")"`;

exports[`valueToSource 13`] = `"undefined"`;

exports[`valueToSource 14`] = `"new Date(\\"2020-12-11T10:03:18.520Z\\")"`;
